packages/backend-function/src/function_client_config_generator.ts

```diff
@@ -0,0 +1,128 @@
+} from "aws-amplify/auth";
+
+class CustomCredentialsProvider implements CredentialsAndIdentityIdProvider {
+  env: Env;
+  constructor(env: Env) {
+    this.env = env;
+  }
+
+  async getCredentialsAndIdentityId(): Promise<
+    CredentialsAndIdentityId | undefined
+  > {
+    const credentials: CredentialsAndIdentityId = {
+      credentials: {
+        accessKeyId: this.env.AWS_ACCESS_KEY_ID,
+        secretAccessKey: this.env.AWS_SECRET_ACCESS_KEY,
+        sessionToken: this.env.AWS_SESSION_TOKEN,
+      },
+    };
+    return credentials;
+  }
+  clearCredentialsAndIdentityId(): void {}
+}
+
+type Env = {
+  AWS_ACCESS_KEY_ID: string;
+  AWS_SECRET_ACCESS_KEY: string;
+  AWS_SESSION_TOKEN: string;
+  AMPLIFY_DATA_GRAPHQL_ENDPOINT?: string;
+  AWS_REGION: string;
+};
+
+/**
+ * Configures the amplify library for lambda execution with data-schema client configuration
+ *
+ * @param env The env vars provided for lambda execution
+ * @param modelIntrospection The MIS generated for the schema
+ * @returns the two named inputs to use with \`Amplify.configure(resourceConfig, libraryOptions)\`
+ */
+function lambdaConfig(env: Env, modelIntrospection: object) {
+  const customCredentialsProvider = new CustomCredentialsProvider(env);
+
+  return {
+    resourceConfig: {
+      API: {
+        GraphQL: {
+          endpoint: env.AMPLIFY_DATA_GRAPHQL_ENDPOINT ?? "",
+          region: env.AWS_REGION,
+          defaultAuthMode: "iam" as const,
+          modelIntrospection: modelIntrospection as any,
+        },
+      },
+    },
+    libraryOptions: {
+      Auth: {
+        credentialsProvider: customCredentialsProvider,
+      },
+    },
+  };
+};
+const { resourceConfig, libraryOptions } = lambdaConfig(
+  env,
+  modelIntrospection
+);
+
+export { resourceConfig, libraryOptions };
+`;
+
+/**
+ * Generates a client config for environment variables
+ */
+export class FunctionClientConfigGenerator {
+  private readonly header =
+    '// This file is auto-generated by Amplify. Edits will be overwritten.';
+
+  private clientConfigFilePath: string;
+
+  private indentation: string = '  ';
+
+  /**
+   * Initialize client config file name and location
+   */
+  constructor(private readonly functionName: string) {
+    this.clientConfigFilePath = `${process.cwd()}/.amplify/generated/client-config/${
+      this.functionName
+    }.ts`;
+  }
+  /**
+   * Generate a client configuration
+   */
+  generateClientConfig = () => {
+    this.writeFile(
+      [this.header, lambdaConfigHelper(this.functionName)].join(EOL)
+    );
+  };
+
+  /**
+   * Check if the script content has a client-config import
+   * @param scriptContent script content to check
+   * @returns true if the script content has a client-config import, false otherwise
+   */
+  usedBy(scriptContent: string) {
+    // Matches an import of client-config
+    const clientConfigImportRegex = /import.*from.*client-config/g;
+    return clientConfigImportRegex.test(scriptContent);
+  }
```

Comment:
We've talked about only generating these files if the customer is using them. This does that, but leaves me with concerns.
1. What happens if the customer has multiple files where the handler doesn't import the file we're looking for?
2. What happens if the customer us using a different code consumption pattern such as require?

I considered doing something more complicated like loading the script AST, but settled on starting with this and following up with a discussion.

Reply:
And they can also use good old `require` :( .

I left comment somewhere else to use access grants for this purpose.

If we keep this approach one option is to run bundling twice and inspect the bundled file. then regenerate.

Reply:
Talking with @josefaidt yesterday, the impact of generating these files for every function is minimal. I will experiment with resource permission based filtering this morning. This is discussed in a couple places and I'll use this thread for followups on this topic.

Reply:
> the impact of generating these files for every function is minimal

performance wise perhaps yes. But it creates unwanted link in resource graph making the "cycling dependency problem" more likely to happen. This is not "minimal".

Reply:
I don't have context on the cycling dep problem. It may be related to the error we're chasing right now. I expect to followup on this in another update.

I think the goal is to use the resource auth rules to limit which functions we build files for, however these permissions are causing the current problem, so we need to overcome that before I can test any solution that would constrain based on these auth rules.